internal/cri/server/container_stats_list.go

```diff
@@ -440,6 +469,17 @@ func getAvailableBytesV2(memory *cg2.MemoryStat, workingSetBytes uint64) uint64
 	return 0
 }
 
+func getPidCount(stats interface{}) (uint64, error) {
+	switch metrics := stats.(type) {
+	case *cg1.Metrics:
+		return metrics.GetPids().GetCurrent(), nil
+	case *cg2.Metrics:
+		return metrics.GetPids().GetCurrent(), nil
+	default:
+		return 0, fmt.Errorf("unexpected metrics type: %T from %s", metrics, reflect.TypeOf(metrics).Elem().PkgPath())
+	}
```

Comment:
Looking at the code further up in linuxContainerMetrics (where this function is called), it looks like that function accepts `wstats.Statistics`, but here we produce an error for it.


https://github.com/containerd/containerd/blob/faf06a3ae29f194d413239d544bf986a9c09fda8/internal/cri/server/container_stats_list.go#L362-L370

So probably we should either add the type here (and ignore), or change that switch above to return an error if we can't accept it.

I'm also wondering if we should roll up the unmarshaling inside those branches at the cost of duplicating 3 lines of code, so that we don't have to do the type-assertions multiple times;

```go
if err := typeurl.UnmarshalTo(stats.Data, data); err != nil {
	return containerStats{}, fmt.Errorf("failed to extract container metrics: %w", err)
}
```

In that case we can inline the code to get the number of pids;

```go
var data interface{}
var pids int
switch {
case typeurl.Is(stats.Data, (*cg1.Metrics)(nil)):
    data = &cg1.Metrics{}
    if err := typeurl.UnmarshalTo(stats.Data, data); err != nil {
        return nil, fmt.Errorf("failed to extract container metrics: %w", err)
    }
    pids = metrics.GetPids().GetCurrent()
case typeurl.Is(stats.Data, (*cg2.Metrics)(nil)):
    data = &cg2.Metrics{}
    if err := typeurl.UnmarshalTo(stats.Data, data); err != nil {
        return nil, fmt.Errorf("failed to extract container metrics: %w", err)
    }
    pids = metrics.GetPids().GetCurrent()
case typeurl.Is(stats.Data, (*wstats.Statistics)(nil)):
    // FIXME: should this be an error on Linux??
    data = &wstats.Statistics{}
    if err := typeurl.UnmarshalTo(stats.Data, data); err != nil {
        return nil, fmt.Errorf("failed to extract container metrics: %w", err)
    }
default:
    return nil, errors.New("cannot convert metric data to cgroups.Metrics or windows.Statistics")
}
```

Reply:
```
case typeurl.Is(stats.Data, (*wstats.Statistics)(nil)): 
 	data = &wstats.Statistics{} 
```

Actually, I am not sure if this code is dead as I think this is for windows platform. At least `internal/cri/server/sandbox_stats_linux.go` should never care about `wstats.Statistics` since it assumes Linux.

But anyway, I think it's safe to have pids as 0 without returning an error here.

Reply:
well, I had a second look and found `cpuContainerStats` doesn't accept `wstats.Statistics` at all https://github.com/containerd/containerd/blob/5258d9f38bfa0dc7bc8fb9e0d2d17b70fd540f13/internal/cri/server/container_stats_list.go#L475-L498

It should be safe to return an error early here.